File: Day328_2025/Q328_GFG.py
class DSU:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def unite(self, a, b):
        a = self.find(a)
        b = self.find(b)
        if a == b:
            return False

        if self.rank[a] < self.rank[b]:
            a, b = b, a

        self.parent[b] = a
        if self.rank[a] == self.rank[b]:
            self.rank[a] += 1

        return True


class Solution:
    def secondMST(self, V, edges):
        E = len(edges)

        # Convert edges to list of (u, v, w)
        e = [(u, v, w) for (u, v, w) in edges]

        # Sort by weight
        e.sort(key=lambda x: x[2])

        d1 = DSU(V)
        mst_index = []
        mstWeight = 0

        # First MST
        for i in range(E):
            u, v, w = e[i]
            if d1.unite(u, v):
                mstWeight += w
                mst_index.append(i)

        if len(mst_index) != V - 1:
            return -1  # No MST possible

        answer = float('inf')

        # Try skipping each MST edge
        for skip in mst_index:
            d2 = DSU(V)
            curWeight = 0
            used = 0

            for i in range(E):
                if i == skip:
                    continue
                u, v, w = e[i]
                if d2.unite(u, v):
                    curWeight += w
                    used += 1
                    if curWeight >= answer:
                        break

            if used == V - 1 and curWeight > mstWeight:
                answer = min(answer, curWeight)

        return -1 if answer == float('inf') else answer

File: Day328_2025/Q328_GFG.java
class DSU {
    int[] parent, rank;

    DSU(int n) {
        parent = new int[n];
        rank = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;
    }

    int find(int x) {
        if (parent[x] == x) return x;
        return parent[x] = find(parent[x]);
    }

    boolean unite(int a, int b) {
        a = find(a);
        b = find(b);
        if (a == b) return false;

        if (rank[a] < rank[b]) {
            int t = a; a = b; b = t;
        }
        parent[b] = a;
        if (rank[a] == rank[b]) rank[a]++;
        return true;
    }
}

class Q328_GFG {
    public int secondMST(int V, int[][] edges) {

        int E = edges.length;

        // Convert into list of (u, v, w)
        int[][] e = new int[E][3];
        for (int i = 0; i < E; i++) {
            e[i][0] = edges[i][0];
            e[i][1] = edges[i][1];
            e[i][2] = edges[i][2];
        }

        // Sort by weight
        java.util.Arrays.sort(e, (a, b) -> a[2] - b[2]);

        DSU d1 = new DSU(V);
        java.util.List<Integer> mstIndex = new java.util.ArrayList<>();
        int mstWeight = 0;

        // First MST
        for (int i = 0; i < E; i++) {
            if (d1.unite(e[i][0], e[i][1])) {
                mstWeight += e[i][2];
                mstIndex.add(i);
            }
        }

        if (mstIndex.size() != V - 1)
            return -1; // no MST possible

        int answer = Integer.MAX_VALUE;

        // Try skipping each MST edge
        for (int skip : mstIndex) {
            DSU d2 = new DSU(V);
            int curWeight = 0;
            int used = 0;

            for (int i = 0; i < E; i++) {
                if (i == skip) continue;
                if (d2.unite(e[i][0], e[i][1])) {
                    curWeight += e[i][2];
                    used++;

                    if (curWeight >= answer) break;
                }
            }

            if (used == V - 1 && curWeight > mstWeight)
                answer = Math.min(answer, curWeight);
        }

        return answer == Integer.MAX_VALUE ? -1 : answer;
    }
}
